### ISSUE TITLE
Compiling OCaml with llvm15 as the c compiler trigger compiler warning for file interp.c

### ISSUE THREAD

**Issue by richard-zhang** [ocaml/ocaml#11759]
Trying to compile OCaml with llvm15 where deprecated-non-prototype is enabled by default gives the following error for interp.c file. Could you gives some hints how to fix the warning for interp.c? Or shall we add `-Wnodeprecated-non-prototype` to c_warnings flags? 

## Reproduce
* command:
```
/opt/homebrew/opt/llvm/bin/clang -c -O2 -fno-strict-aliasing -fwrapv -pthread -g -Wall -Werror -fno-common  -I./runtime  -D_FILE_OFFSET_BITS=64  -DCAMLDLLIMPORT= -DIN_CAML_RUNTIME -I/opt/homebrew/opt/llvm/include -o runtime/interp.b.o runtime/interp.c
```
* error:
```
runtime/interp.c:1037:28: error: passing arguments to a function without a prototype is deprecated in all versions of C and is not supported in C2x [-Werror,-Wdeprecated-non-prototype]
      accu = Primitive(*pc)(accu);
```

**Comment by nojb** [comment]
This warning cannot be easily fixed because the value denoted by the macro `Primitive(*pc)` can be of any arity. Dedicated macros for each arity would need to be introduced.

> Or shall we add `-Wnodeprecated-non-prototype` to c_warnings flags?

I don't know what `c_warnings` means, but this sounds like a good fix you can do on your side for the time being.

**Comment by richard-zhang** [comment]
I'm referring to add an additional flag here https://github.com/ocaml/ocaml/blob/trunk/configure.ac#L690
from `-Wall` to `-Wall -Wnodeprecated-non-prototype`

**Comment by richard-zhang** [comment]
Currently, this won't be an issue. But once Apple upgrade the system cc `/usr/bin/cc` from 14 to 15, we might see this errors more frequent for Mac users.

**Comment by dra27** [comment]
@MisterDA has been doing some work around this which may help. Otherwise, I agree adding `-Wnodeprecated-non-prototype` would be good, but I would do this specifically for `runtime/interp.c` rather than for the entire build system. Not sure exactly where we are with per-file CFLAGS, though, @shindere? (the change would be around [here](https://github.com/ocaml/ocaml/blob/trunk/Makefile#L867-L892) either to add the flag just to C files in runtime/ or, preferably, to all the files with `interp.` as a stem)

**Comment by shindere** [comment]
David Allsopp (2022/11/27 07:12 -0800):
> @MisterDA has been doing some work around this which may help.
> Otherwise, I agree adding `-Wnodeprecated-non-prototype` would be
> good, but I would do this specifically for `runtime/interp.c` rather
> than for the entire build system. Not sure exactly where we are with
> per-file CFLAGS, though, @shindere? (the change would be around
> [here](https://github.com/ocaml/ocaml/blob/trunk/Makefile#L867-L892)
> either to add the flag just to C files in runtime/ or, preferably, to
> all the files with `interp.` as a stem)

I think, either we make the build macros more sophisticated so that they
can use per-file variables, e.g. the comand that builds `interp.o`
would read `$(interp_CFLAGS)` etc., or, for an easier and more ad hoc
approach, adding a pattern rule that just adds the appropriate flags to
`OC_CFLAGS` for the files that need it can work, too. If there are
not too many things of this kind I'd go for the second approach
I think, and switch to the first one only if it turns out that there
is a real need for it, by real I mean, e.g.¿ in several places.

**Comment by xavierleroy** [comment]
> This warning cannot be easily fixed because the value denoted by the macro Primitive(*pc) can be of any arity. Dedicated macros for each arity would need to be introduced.

That's the easy part.  Not having to deal with primitive arities make it much easier to autogenerate the runtime/prims.c file.  I'd like to have proper support for primitive arities, since K&R-style unprototyped C functions are a thing of the past that must disappear, and I think I know how to do this.  However, generating runtime/prims.c will take some more work,.  I'm currently blocked from using a decent scripting language in the OCaml build system and don't see myself doing this just with sed and awk.

**Comment by dra27** [comment]
We should have a look at what @MisterDA has been putting together (which was still done using `sed`). However, another option - especially if we're generating a "nicer" prims.c - might be to be committing it (as we do with the parser already and since 5.0 as we now do with runtime/caml/sizeclasses.h). It could still be possible to develop the compiler with runtime/prims.c being regenerated (not unlike the computation of C dependencies, which is turned off on releases)

**Comment by edwintorok** [comment]
FWIW I had to turn off one more warning to get OCaml 5.0 to build with `clang`:  `-Wno-deprecated-non-prototype -Wno-unused-but-set-variable`, however the latter is already fixed on trunk, so I opened a PR here to backport it to 5.0 too: https://github.com/ocaml/ocaml/pull/11845

**Comment by dra27** [comment]
@edwintorok - just to clarify, is the problem building the 5.0 dev branch (where `-Werror` is on by default) or building a release tarball/opam switch (where warnings should definitely not be errors)?

**Comment by edwintorok** [comment]
@dra27 I was building the 5.0 branch from git. Having Werror off in release tarballs sounds like a good choice.

**Comment by dra27** [comment]
I think this has been fixed with #11861?

**Comment by shindere** [comment]
I know the issue is currently closed, but would still be interested in
@MisterDA 's opinion on it. What do you think, Antonin?
